Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+#include <AzToolsFramework/ComponentMode/EditorComponentModeBus.h>
+#include <AzToolsFramework/Entity/EditorEntityHelpers.h>
+#include <AzToolsFramework/ToolsComponents/EditorComponentBase.h>
+#include <AzToolsFramework/ViewportSelection/EditorTransformComponentSelection.h>
+#include <AzToolsFramework/ViewportUi/ViewportUiManager.h>
+
+namespace AzToolsFramework::ComponentModeFramework
+
+{
+    // ComponentData constructor to fill in component info
+    ComponentData::ComponentData(AZ::EntityComponentIdPair pairId)
+        : m_pairId(pairId)
+    {
+        AZ::ComponentApplicationBus::BroadcastResult(
+            m_entity, &AZ::ComponentApplicationBus::Events::FindEntity, AZ::EntityId(pairId.GetEntityId()));
+        AZ::ComponentId componentId = pairId.GetComponentId();
+
+        m_component = m_entity->FindComponent(componentId);
+
+        m_componentName = AzToolsFramework::GetFriendlyComponentName(m_component);
+
+        AzToolsFramework::EditorRequestBus::BroadcastResult(
+            m_iconPath,
+            &AzToolsFramework::EditorRequestBus::Events::GetComponentEditorIcon,
+            m_component->GetUnderlyingComponentType(),
+            m_component);
+    }
+
+    ComponentModeSwitcher::ComponentModeSwitcher()
+    {
+        ViewportEditorModeNotificationsBus::Handler::BusConnect(GetEntityContextId());
+        EntityCompositionNotificationBus::Handler::BusConnect();
+        ToolsApplicationNotificationBus::Handler::BusConnect();
+
+
+        // Create the switcher
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            m_switcherId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcher,
+            ViewportUi::Alignment::TopLeft);
+
+        // Initial Transform button
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            m_transformButtonId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcherButton,
+            m_switcherId,
+            "../../../../Assets/Editor/Icons/Components/Transform.svg",
+            "Transform");
+
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherActiveButton,
+            m_switcherId,
+            m_transformButtonId);
+
+        // Handler
+        auto handlerFunc = [this](ViewportUi::ButtonId buttonId)
+        {
+            ViewportUi::ViewportUiRequestBus::Event(
+                ViewportUi::DefaultViewportId,
+                &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherActiveButton,
+                m_switcherId,
+                buttonId);
+
+            ActivateComponentMode(buttonId);
+        };
+
+        m_handler = AZ::Event<ViewportUi::ButtonId>::Handler(handlerFunc);
+        // Registers the SwitcherEventHandler which is defined in ComponentModeSwitcher
```

Comment:
nit: This is really subtle but in source files (code changes) comments tend to use the active voice, so they describe what the code does immediately. So this would be...

```
// register the SwitcherEventHandler
```

When commenting an API though, it's nice to use the passive 'Registers...' or 'Removes...' instead of 'Register' or 'Remove'. It's a subtle difference but good to try and stick to for consistency.